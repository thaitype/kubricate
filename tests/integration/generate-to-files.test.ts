import fs from 'node:fs/promises';
import path from 'node:path';

import { rimraf } from 'rimraf';
import { afterEach, describe, expect, it } from 'vitest';

import { executeKubricate } from '../helpers/execute-kubricate';

const rootDir = path.resolve(__dirname, '..');
const fixturesRoot = path.join(rootDir, 'fixtures');

const scenarios = [
  { name: 'Output Mode: Stack', fixture: 'generate-output-stack' },
  { name: 'Output Mode: Flat', fixture: 'generate-output-flat' },
  { name: 'Output Mode: Resource', fixture: 'generate-output-resource' },
  { name: 'Output Mode: Flat With Secret', fixture: 'generate-output-with-secret' },
] as const;

async function snapshotDirectory(dir: string, fixturePrefix: string) {
  const entries = await fs.readdir(dir, { withFileTypes: true });

  await Promise.all(
    entries.map(async entry => {
      const fullPath = path.join(dir, entry.name);
      const relativePath = path.join(fixturePrefix, entry.name);

      if (entry.isDirectory()) {
        await snapshotDirectory(fullPath, relativePath);
      } else {
        const content = await fs.readFile(fullPath, 'utf-8');
        expect(content).toMatchSnapshot(relativePath);
      }
    })
  );
}

describe.each(scenarios)('CLI Integration ($name)', ({ fixture }) => {
  const fixturesDir = path.join(fixturesRoot, fixture);
  const outputDir = 'output';
  const outputFixtureDir = path.join(fixturesDir, outputDir);

  afterEach(async () => {
    await rimraf(outputFixtureDir);
  });

  it('should generate expected files', async () => {
    const args = ['generate', '--root', fixturesDir];
    const { stdout, exitCode } = await executeKubricate(args, { reject: false });

    expect(exitCode).toBe(0);
    expect(stdout).toContain('Generating stacks');

    await snapshotDirectory(outputFixtureDir, `${fixture}/${outputDir}`);
  });
});
