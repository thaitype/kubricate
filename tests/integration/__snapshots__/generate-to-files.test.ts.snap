// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI Integration ('Output Mode: Flat With Secret') > should generate expected files > generate-output-with-secret/output/stacks.yml 1`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: namespace
    kubricate.thaitype.dev/resource-id: namespace
  annotations:
    kubricate.thaitype.dev/stack-name: NamespaceStack
    kubricate.thaitype.dev/resource-hash: defd4c342aba95d18a5e0ed81ab55ed132770c5945f11b39c927c4b6d003a6a2
---

apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-app
  namespace: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: frontend
    kubricate.thaitype.dev/resource-id: deployment
  annotations:
    kubricate.thaitype.dev/stack-name: SimpleAppStack
    kubricate.thaitype.dev/resource-hash: c9325f337dfc06f0a2245d994a63be2f0b13900d0bb41fa19a4a642451cd5c59
spec:
  replicas: 1
  selector:
    matchLabels:
      app: my-app
  template:
    metadata:
      labels:
        app: my-app
    spec:
      containers:
        - image: nginx
          name: my-app
          ports:
            - containerPort: 80
          env:
            - name: API_KEY
              valueFrom:
                secretKeyRef:
                  name: secret-application
                  key: my_app_key
            - name: API_KEY_2
              valueFrom:
                secretKeyRef:
                  name: secret-application
                  key: my_app_key_2
      imagePullSecrets:
        - name: secret-application-provider
---

apiVersion: v1
kind: Service
metadata:
  name: my-app
  namespace: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: frontend
    kubricate.thaitype.dev/resource-id: service
  annotations:
    kubricate.thaitype.dev/stack-name: SimpleAppStack
    kubricate.thaitype.dev/resource-hash: fa9a081478553a38803e92417d6542fb29906a284784ece83aa6c5e5f87138d4
spec:
  selector:
    app: my-app
  type: ClusterIP
  ports:
    - port: 80
      targetPort: 80
---
"
`;

exports[`CLI Integration ('Output Mode: Flat') > should generate expected files > generate-output-flat/output/stacks.yml 1`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: namespace
    kubricate.thaitype.dev/resource-id: namespace
  annotations:
    kubricate.thaitype.dev/stack-name: NamespaceStack
    kubricate.thaitype.dev/resource-hash: defd4c342aba95d18a5e0ed81ab55ed132770c5945f11b39c927c4b6d003a6a2
---

apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-app
  namespace: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: frontend
    kubricate.thaitype.dev/resource-id: deployment
  annotations:
    kubricate.thaitype.dev/stack-name: SimpleAppStack
    kubricate.thaitype.dev/resource-hash: 956f8bab85981ce685e2cce42949131bad2f92accf4832a66f56e9da81d4a8a0
spec:
  replicas: 1
  selector:
    matchLabels:
      app: my-app
  template:
    metadata:
      labels:
        app: my-app
    spec:
      containers:
        - image: nginx
          name: my-app
          ports:
            - containerPort: 80
---

apiVersion: v1
kind: Service
metadata:
  name: my-app
  namespace: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: frontend
    kubricate.thaitype.dev/resource-id: service
  annotations:
    kubricate.thaitype.dev/stack-name: SimpleAppStack
    kubricate.thaitype.dev/resource-hash: fa9a081478553a38803e92417d6542fb29906a284784ece83aa6c5e5f87138d4
spec:
  selector:
    app: my-app
  type: ClusterIP
  ports:
    - port: 80
      targetPort: 80
---
"
`;

exports[`CLI Integration ('Output Mode: Resource') > should generate expected files > generate-output-resource/output/frontend/Deployment_deployment.yml 1`] = `
"apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-app
  namespace: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: frontend
    kubricate.thaitype.dev/resource-id: deployment
  annotations:
    kubricate.thaitype.dev/stack-name: SimpleAppStack
    kubricate.thaitype.dev/resource-hash: 956f8bab85981ce685e2cce42949131bad2f92accf4832a66f56e9da81d4a8a0
spec:
  replicas: 1
  selector:
    matchLabels:
      app: my-app
  template:
    metadata:
      labels:
        app: my-app
    spec:
      containers:
        - image: nginx
          name: my-app
          ports:
            - containerPort: 80
---
"
`;

exports[`CLI Integration ('Output Mode: Resource') > should generate expected files > generate-output-resource/output/frontend/Service_service.yml 1`] = `
"apiVersion: v1
kind: Service
metadata:
  name: my-app
  namespace: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: frontend
    kubricate.thaitype.dev/resource-id: service
  annotations:
    kubricate.thaitype.dev/stack-name: SimpleAppStack
    kubricate.thaitype.dev/resource-hash: fa9a081478553a38803e92417d6542fb29906a284784ece83aa6c5e5f87138d4
spec:
  selector:
    app: my-app
  type: ClusterIP
  ports:
    - port: 80
      targetPort: 80
---
"
`;

exports[`CLI Integration ('Output Mode: Resource') > should generate expected files > generate-output-resource/output/namespace/Namespace_namespace.yml 1`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: namespace
    kubricate.thaitype.dev/resource-id: namespace
  annotations:
    kubricate.thaitype.dev/stack-name: NamespaceStack
    kubricate.thaitype.dev/resource-hash: defd4c342aba95d18a5e0ed81ab55ed132770c5945f11b39c927c4b6d003a6a2
---
"
`;

exports[`CLI Integration ('Output Mode: Stack') > should generate expected files > generate-output-stack/output/frontend.yml 1`] = `
"apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-app
  namespace: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: frontend
    kubricate.thaitype.dev/resource-id: deployment
  annotations:
    kubricate.thaitype.dev/stack-name: SimpleAppStack
    kubricate.thaitype.dev/resource-hash: 956f8bab85981ce685e2cce42949131bad2f92accf4832a66f56e9da81d4a8a0
spec:
  replicas: 1
  selector:
    matchLabels:
      app: my-app
  template:
    metadata:
      labels:
        app: my-app
    spec:
      containers:
        - image: nginx
          name: my-app
          ports:
            - containerPort: 80
---

apiVersion: v1
kind: Service
metadata:
  name: my-app
  namespace: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: frontend
    kubricate.thaitype.dev/resource-id: service
  annotations:
    kubricate.thaitype.dev/stack-name: SimpleAppStack
    kubricate.thaitype.dev/resource-hash: fa9a081478553a38803e92417d6542fb29906a284784ece83aa6c5e5f87138d4
spec:
  selector:
    app: my-app
  type: ClusterIP
  ports:
    - port: 80
      targetPort: 80
---
"
`;

exports[`CLI Integration ('Output Mode: Stack') > should generate expected files > generate-output-stack/output/namespace.yml 1`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name: my-namespace
  labels:
    kubricate.thaitype.dev: "true"
    kubricate.thaitype.dev/stack-id: namespace
    kubricate.thaitype.dev/resource-id: namespace
  annotations:
    kubricate.thaitype.dev/stack-name: NamespaceStack
    kubricate.thaitype.dev/resource-hash: defd4c342aba95d18a5e0ed81ab55ed132770c5945f11b39c927c4b6d003a6a2
---
"
`;
